public/app/plugins/datasource/influxdb/datasource.ts

```diff
@@ -381,67 +384,67 @@ export default class InfluxDatasource extends DataSourceWithBackend<InfluxQuery,
       return [];
     }
 
-    let expandedQueries = queries;
-    if (queries && queries.length > 0) {
-      expandedQueries = queries.map((query) => {
-        if (this.isFlux) {
-          return {
-            ...query,
-            datasource: this.getRef(),
-            query: this.templateSrv.replace(query.query ?? '', scopedVars, 'regex'),
-          };
-        }
-
-        const expandedQuery = {
+    return queries.map((query) => {
+      if (this.isFlux) {
+        return {
           ...query,
           datasource: this.getRef(),
-          measurement: this.templateSrv.replace(query.measurement ?? '', scopedVars, 'regex'),
-          policy: this.templateSrv.replace(query.policy ?? '', scopedVars, 'regex'),
-          limit: this.templateSrv.replace(query.limit?.toString() ?? '', scopedVars, 'regex'),
-          slimit: this.templateSrv.replace(query.slimit?.toString() ?? '', scopedVars, 'regex'),
-          tz: this.templateSrv.replace(query.tz ?? '', scopedVars),
+          query: this.templateSrv.replace(query.query ?? '', scopedVars), // The raw query text
         };
+      }
 
-        if (query.rawQuery) {
-          expandedQuery.query = this.templateSrv.replace(query.query ?? '', scopedVars, 'regex');
-        }
+      return {
+        ...query,
+        datasource: this.getRef(),
+        ...this.applyVariables(query, scopedVars, scopedVars),
+      };
+    });
+  }
 
-        if (query.groupBy) {
-          expandedQuery.groupBy = query.groupBy.map((groupBy) => {
-            return {
-              ...groupBy,
-              params: groupBy.params?.map((param) => {
-                return this.templateSrv.replace(param.toString(), undefined, 'regex');
-              }),
-            };
-          });
-        }
+  applyVariables(query: InfluxQuery, scopedVars: ScopedVars, rest: ScopedVars) {
+    if (query.groupBy) {
+      query.groupBy = query.groupBy.map((groupBy) => {
+        return {
```

Comment:
I have seen problems before when mutating the query direcly like this. But can't remember what. Can we use the original method like `const expandedQuery = { ...query }` and then return the copy?

@gabor

Reply:
i would also recommend not mutating the query (i don't have any specific problem in mind that could be caused by this, but in general we should not mutate what we do not create).

we have https://immerjs.github.io/immer/ in our `package.json`, i recommend using it here if the `{...query}` approach gets too long.

Reply:
Thanks for the tip.. pushed changes.